packages/vite/src/node/server/index.ts

```diff
@@ -541,17 +541,22 @@ export async function createServer(
     }
   }
 
+  let isOptimized = false
   if (!middlewareMode && httpServer) {
     // overwrite listen to run optimizer before server start
     const listen = httpServer.listen.bind(httpServer)
     httpServer.listen = (async (port: number, ...args: any[]) => {
+      if (isOptimized) {
+        return listen(port, ...args)
+      }
       try {
         await container.buildStart({})
         await runOptimize()
       } catch (e) {
         httpServer.emit('error', e)
         return
       }
+      isOptimized = true
       return listen(port, ...args)
     }) as any
```

Comment:
I'm having second thoughts on restoring the original `listen` as it may confuse people while debugging. I think your approach with the flag is better in this case to avoid surprises. The only thing is that re-organizing the flag check would have helped me when reviewing the code. What do you think?
```suggestion
  if (!middlewareMode && httpServer) {
    let isOptimized = false
    // overwrite listen to run optimizer before server start
    const listen = httpServer.listen.bind(httpServer)
    httpServer.listen = (async (port: number, ...args: any[]) => {
      if (!isOptimized) {
        try {
          await container.buildStart({})
          await runOptimize()
          isOptimized = true
        } catch (e) {
          httpServer.emit('error', e)
          return
        }
      }
      return listen(port, ...args)
    }) as any
```

Reply:
It's good! Thank you for your review.